### PR TITLE
enable precompilation

### LINKED ISSUE #30: Tests fail when package is precompiled
```
julia> versioninfo()
Julia Version 0.4.0-dev+6574
Commit ebd7230 (2015-08-10 17:56 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin14.4.0)
  CPU: Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3

julia> Pkg.test("LightXML")
INFO: Computing test dependencies for LightXML...
INFO: No packages to install, update or remove
INFO: Testing LightXML
running parse.jl ...
ERROR: LoadError: LoadError: UndefRefError: access to undefined reference
 in parse_string at /Users/seth/.julia/v0.4/LightXML/src/document.jl:94
 in include at /usr/local/julia-latest/lib/julia/sys.dylib
 in include_from_node1 at /usr/local/julia-latest/lib/julia/sys.dylib
 in anonymous at no file:6
 in include at /usr/local/julia-latest/lib/julia/sys.dylib
 in include_from_node1 at /usr/local/julia-latest/lib/julia/sys.dylib
 in process_options at /usr/local/julia-latest/lib/julia/sys.dylib
 in _start at /usr/local/julia-latest/lib/julia/sys.dylib
while loading /Users/seth/.julia/v0.4/LightXML/test/parse.jl, in expression starting on line 24
while loading /Users/seth/.julia/v0.4/LightXML/test/runtests.jl, in expression starting on line 3
========================================================[ ERROR: LightXML ]=========================================================

failed process: Process(`/usr/local/julia-latest/bin/julia --check-bounds=yes --code-coverage=none --color=yes /Users/seth/.julia/v0.4/LightXML/test/runtests.jl`, ProcessExited(1)) [1]

====================================================================================================================================
INFO: No packages to install, update or remove
ERROR: LightXML had test errors
 in error at /usr/local/julia-latest/lib/julia/sys.dylib
 in test at pkg/entry.jl:746
 in anonymous at pkg/dir.jl:31
 in cd at file.jl:22
 in cd at pkg/dir.jl:31
 in test at pkg.jl:71
```

(I noticed this in LightGraphs, which uses LightXML as a dependency now. It's failing on `parse_file()` (line 88) there.)

### PR DESCRIPTION
**Pull request by stevengj**
This fixes #30 to enable precompilation.  While auditing the code, I implemented several small cosmetic improvements and some improved 0.4 compatibility (using `Cstring`, `String` to `AbstractString`, etc.).

### CODE REVIEW

**Comment by tkelman** [comment]
Cool, not too bad. I don't know what the coverage here is like, but this also strikes me as worth a minor bump just in case any reverse-deps somehow hit trouble.

**Comment by stevengj** [comment]
@tkelman, could you elaborate?  Since the (exported) API hasn't changed, how could reverse-deps have a problem with this?  (Non-precompiled modules can import precompiled modules.)

File: src/LightXML.jl [hunk 1]
```diff
@@ -22,12 +26,21 @@ module LightXML
...
     parse_file, parse_string, save_file, set_root, create_root
 
-    using Compat
-    include("clib.jl")
-    include("errors.jl")
+typealias Xchar UInt8
+typealias Xstr Ptr{Xchar}
+
+# opaque pointer type (do not dereference!) corresponding to xmlBufferPtr in C
+immutable xmlBuffer; end
```
Comment (yuyichao) [inline src/LightXML.jl]: The semicolon here is not necessary. Doesn't matter, just can't resist.

File: src/clib.jl [hunk 2]
```diff
@@ -1,69 +0,0 @@
...
-
-# functions for documents
-
-@lx2func xmlDocGetRootElement
-@lx2func xmlDocSetRootElement
-@lx2func xmlNewDoc
-@lx2func xmlFreeDoc
-@lx2func xmlParseFile
-@lx2func xmlParseMemory
-@lx2func xmlDocDumpMemoryEnc
```
Comment (tkelman) [inline src/clib.jl]: Curious, this and `xmlSaveFileEnc` weren't bound anywhere? Easy enough to use the normal way if anyone needs them.

File: src/clib.jl [hunk 3]
```diff
@@ -1,69 +0,0 @@
...
-
-typealias Xchar Uint8
-typealias Xstr Ptr{Xchar}
-typealias Xptr Ptr{Void}
-
-# supporting functions
-
-#
-# After tests, it seems that free in libc instead of xmlFree
-# should be used here
```
Comment (yuyichao) [inline src/clib.jl]: This looks like a comment I'd like to keep for future references. Ref https://github.com/JuliaLang/LightXML.jl/commit/b33d030cb052f05d528a7355f2237471103a394c @lindahua

**Comment by tkelman** [comment]
The exported API hasn't changed, but the internal implementation changes a bit. I doubt anyone will notice but I'm conservative until thorough testing shows otherwise. If it takes nontrivial code changes to be able to enable precompilation, then I think it's worth marking that in a way reverse-deps could differentiate between in a fine-grained way on the off chance it ends up necessary.

File: src/clib.jl (hunk 3 above)
Comment (stevengj) [inline src/clib.jl]: fixed

**Comment by stevengj** [comment]
Okay.  Version numbers are cheap, anyway.

File: src/LightXML.jl [hunk 4]
```diff
@@ -22,12 +26,23 @@ module LightXML
...
+typealias Xstr Ptr{Xchar}
+
+# opaque pointer type (do not dereference!) corresponding to xmlBufferPtr in C
+immutable xmlBuffer end
+typealias Xptr Ptr{xmlBuffer}
+
+# pre-condition: p is not null
+# (After tests, it seems that free in libc instead of xmlFree
+#  should be used here.)
+_xcopystr(p::Xstr) = (r = bytestring(p); Libc.free(p); r)
```
Comment (stevengj) [inline src/LightXML.jl]: Note that this could use `Base.pointer_to_string` in 0.4 to avoid making a copy, but that's not currently exported.

**Comment by sbromberger** [comment]
Just wanted to thank @stevengj and @tkelman for working on this. This will enable precompilation of LightGraphs.

**Comment by tkelman** [comment]
@stevengj did all the work, I just reviewed it (and so did @yuyichao, thanks). Pleasantly surprised to see it wasn't much.

**Comment by tkelman** [comment]
And since @stevengj hasn't tagged this yet, I went and finished up my tiny feature-addition PR #27 which will make a minor bump even more justifiable. Anyone mind reviewing?

**Pull request merged**